File: src/utils/account.js
// ── Accounts via invite codes ────────────────────────────────────
// There is no signup. An account is provisioned by the owner (a
// private data repo + a fine-grained PAT scoped to it, optionally a
// Gemini key) and handed out as one base64url invite code made by
// scripts/make-invite.js. Redeeming a code just fills the same
// localStorage keys the app has always used — sync and the AI need
// no changes. Revocation = revoking that PAT on GitHub.

import { setApiKey, getApiKey } from './storage.js';
import { setSyncConfig, getSyncConfig } from '../db/sync.js';

const ACCOUNT_KEY = 'coach:account';

export function getAccount() {
  try {
    return JSON.parse(localStorage.getItem(ACCOUNT_KEY)) || null;
  } catch {
    return null;
  }
}

/** True on a fresh install with no account and no hand-entered setup
 *  (pre-account installs keep working without ever seeing Login). */
export function needsLogin() {
  return !getAccount() && !getApiKey() && !getSyncConfig().token;
}

function b64urlDecode(str) {
  const b64 = str.replace(/-/g, '+').replace(/_/g, '/');
  const bin = atob(b64 + '='.repeat((4 - (b64.length % 4)) % 4));
  const bytes = Uint8Array.from(bin, (c) => c.charCodeAt(0));
  return new TextDecoder().decode(bytes);
}

/** Decode + validate an invite code. Throws a user-readable error. */
export function parseInviteCode(code) {
  const raw = String(code || '').replace(/\s+/g, '');
  if (!raw) throw new Error('Paste the invite code you were sent.');
  let acct;
  try {
    acct = JSON.parse(b64urlDecode(raw));
  } catch {
    throw new Error("That doesn't look like a COACH invite code — check you copied all of it.");
  }
  if (!acct?.name || !acct?.repo || !acct?.ghToken) {
    throw new Error('Invite code is incomplete — ask for a new one.');
  }
  return acct;
}

/** Redeem: write the config the rest of the app already reads. */
export function applyAccount(acct) {
  setSyncConfig({ token: acct.ghToken, repo: acct.repo });
  if (acct.geminiKey) setApiKey(acct.geminiKey);
  localStorage.setItem(
    ACCOUNT_KEY,
    JSON.stringify({ name: acct.name, since: new Date().toISOString() })
  );
}

/** Clear every coach:* key and the local DB (cloud copies untouched). */
export async function wipeLocal() {
  for (const k of Object.keys(localStorage)) {
    if (k.startsWith('coach:')) localStorage.removeItem(k);
  }
  await new Promise((resolve) => {
    const req = indexedDB.deleteDatabase('coach-db');
    req.onsuccess = req.onerror = req.onblocked = () => resolve();
  });
}

/** Wipe this device back to the login screen (data stays in the cloud). */
export async function signOut() {
  await wipeLocal();
  window.location.reload();
}

File: ios/CoachApp/CoachApp/Account/Account.swift
import Foundation

/// Invite-code based accounts — ports src/utils/account.js. There's no
/// signup: the owner provisions each account by hand (a private GitHub
/// data repo + a fine-grained PAT scoped to it, optionally a Gemini key)
/// and hands it out as one base64url invite code from
/// scripts/make-invite.js. Redeeming a code just fills the same
/// Keychain/Defaults slots the rest of the app already reads — sync and
/// the AI need no changes. Uses the identical encoding as the web app so
/// existing invite codes work unmodified.
struct InviteCode: Codable {
    var name: String
    var repo: String
    var ghToken: String
    var geminiKey: String?
}

struct StoredAccount: Codable {
    var name: String
    var since: String
}

enum AccountError: LocalizedError {
    case empty
    case malformed
    case incomplete

    var errorDescription: String? {
        switch self {
        case .empty: return "Paste the invite code you were sent."
        case .malformed: return "That doesn't look like a COACH invite code — check you copied all of it."
        case .incomplete: return "Invite code is incomplete — ask for a new one."
        }
    }
}

enum Account {
    private static let accountKey = "account"

    static func current() -> StoredAccount? {
        Defaults.codable(StoredAccount.self, accountKey)
    }

    /// True on a fresh install with no account and no hand-entered setup
    /// (pre-account installs keep working without ever seeing Login).
    static func needsLogin() -> Bool {
        current() == nil && Keychain.apiKey.isEmpty && GitHubSync.config().token.isEmpty
    }

    /// Decode + validate an invite code. Accepts either the bare code or
    /// a full magic link (https://…/#invite=<code>) — the web app only
    /// auto-extracts the code from a link via App.jsx's boot-time hash
    /// parsing (a separate path from the Login screen's paste box, which
    /// expects the bare code same as here); since this app has no
    /// equivalent hash-based entry point, pasting the whole link into
    /// this box needs to work directly instead. Throws a user-readable
    /// error.
    static func parseInviteCode(_ code: String) throws -> InviteCode {
        var raw = code.replacingOccurrences(of: "\\s+", with: "", options: .regularExpression)
        guard !raw.isEmpty else { throw AccountError.empty }
        if let g = Stats.firstMatch(#"[#&]invite=([A-Za-z0-9_-]+)"#, in: raw), g.count > 1, let extracted = g[1] {
            raw = extracted
        }

        var b64 = raw.replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
        let padding = (4 - b64.count % 4) % 4
        b64 += String(repeating: "=", count: padding)

        guard let data = Data(base64Encoded: b64) else { throw AccountError.malformed }
        guard let acct = try? JSONDecoder().decode(InviteCode.self, from: data) else {
            throw AccountError.malformed
        }
        guard !acct.name.isEmpty, !acct.repo.isEmpty, !acct.ghToken.isEmpty else {
            throw AccountError.incomplete
        }
        return acct
    }

    /// Redeem: write the config the rest of the app already reads.
    static func apply(_ acct: InviteCode) {
        GitHubSync.setConfig(token: acct.ghToken, repo: acct.repo)
        if let key = acct.geminiKey, !key.isEmpty {
            Keychain.apiKey = key
        }
        let stored = StoredAccount(name: acct.name, since: ISO8601DateFormatter().string(from: Date()))
        Defaults.setCodable(stored, for: accountKey)
    }

    /// Clears every stored secret/setting and the local backup file
    /// (cloud copies untouched).
    static func wipeLocal() {
        Keychain.removeAll()
        Defaults.removeAll()
        LocalStore.shared.wipe()
    }

    /// Wipes this device back to the login screen (data stays in the cloud).
    static func signOut() {
        wipeLocal()
    }
}
